ibis/pandas/execution/timecontext.py

```diff
@@ -0,0 +1,115 @@
+column of Timestamp type, and named as 'time' in TableExpr.
+See ``execute_database_table_client`` in ``generic.py``.
+And we assume timecontext is passed in as a tuple (begin, end) where begin and
+end are timestamp, or datetime string like "20100101".
+"""
+
+import numpy as np
+import pandas as pd
+
+import ibis.expr.api as ir
+import ibis.expr.operations as ops
+from ibis.pandas.dispatch import (
+    compute_time_context,
+    is_computable_input,
+    post_execute,
+)
+from ibis.pandas.execution import execute
+
+
+@compute_time_context.register(ops.AsOfJoin)
+def adjust_context_asof_join(op, timecontext=None, **kwargs):
+    new_timecontexts = [
+        timecontext for arg in op.inputs if is_computable_input(arg)
+    ]
+    # right table should look back or forward
+    try:
+        begin, end = map(pd.to_datetime, timecontext)
+        tolerance = op.tolerance
+        if tolerance is not None:
+            timedelta = pd.Timedelta(-tolerance.op().right.op().value)
+            if timedelta <= pd.Timedelta(0):
+                new_begin = begin + timedelta
+                new_end = end
+            else:
+                new_begin = begin
+                new_end = end + timedelta
+        # right table is the second node in children
+        new_timecontexts[1] = (new_begin, new_end)
+    except ValueError:
+        pass
+    finally:
+        return new_timecontexts
+
+
+@compute_time_context.register(ops.WindowOp)
+def adjust_context_window(op, timecontext=None, **kwargs):
+    new_timecontexts = [
+        timecontext for arg in op.inputs if is_computable_input(arg)
+    ]
+    # adjust time context by preceding and following
+    try:
+        begin, end = map(pd.to_datetime, timecontext)
+        result = [begin, end]
+        preceding = op.window.preceding
+        following = op.window.following
+        if preceding is not None:
+            if isinstance(preceding, ir.Expr):
+                new_preceding = execute(preceding)
+            else:
+                new_preceding = preceding
+            if new_preceding and not isinstance(
+                new_preceding, (int, np.integer)
+            ):
+                result[0] = begin - new_preceding
+        if following is not None:
+            if isinstance(following, ir.Expr):
+                new_following = execute(following)
+            else:
+                new_following = following
+            if new_following and not isinstance(
+                new_following, (int, np.integer)
+            ):
+                result[1] = end + new_following
+        new_timecontexts[0] = tuple(result)
+        new_timecontexts[1] = tuple(result)
+    except ValueError:
+        pass
+    finally:
+        return new_timecontexts
+
+
+@post_execute.register(ops.WindowOp, pd.Series)
+def post_execute_window(op, data, timecontext=None, **kwargs):
+    # trim data to original context
+    try:
+        begin, end = map(pd.to_datetime, timecontext)
```

Comment:
We should do this kind of conversion at this level. timecontext should be already converted to pd.Timestamp once we are at this level.

Reply:
Removed